### PR TITLE
Complete Plunder kingdom and events with reactive hook infrastructure

### PR DESCRIPTION
**Pull request by jss367**
## Summary

- Implements all 40 Plunder kingdom cards (31 new + 9 already existed) and 7 Plunder events.
- Adds reactive hook infrastructure: `on_discard_from_play`, `on_card_gained`, `on_action_played`, plus overpay support in the buy phase.
- 44 new tests; full suite is **401 passing** (up from 357 baseline).

## New cards

**Round 1 (12)**: Sack of Loot, Jewelled Egg, Buried Treasure, Longship, Sailor, Cabin Boy, Fortune Hunter, Mapmaker, Cutthroat, Figurine, King's Cache, Search

**Round 2 (19)**: Shaman, Secluded Shrine, Siren, Stowaway, Abundance, Crucible, Gondola, Landing Party, Maelstrom, Pendant, Silver Mine, Tools, Enlarge, Frigate, Rope, Cage, Grotto, Quartermaster, Mining Road

## New events

Bury, Avoid, Foray, Peril, Scrounge, Prosper, Invasion (Looting was already present).

## Infrastructure

| Hook / Mechanic | Purpose |
|---|---|
| `on_discard_from_play(state, player)` | Fires when a card moves from in-play/duration to discard during cleanup. Used by Search, Silver Mine, Cutthroat, Mining Road. |
| `on_card_gained(state, owner, gainer, card)` | Generic gain reaction dispatcher across all players' in-play/duration zones. Used by Cutthroat (Loot trigger), Mining Road (Treasure-swap reaction). |
| `on_action_played(state, player, card)` | Fires after each Action card resolves in the action phase. |
| Overpay in `handle_buy_phase` | Calls `ai.choose_overpay_amount` and dispatches `card.on_overpay`. |
| Default AI methods | `choose_overpay_amount`, `choose_card_to_set_aside`, `quartermaster_choice`, `gondola_delay_coins`, `cabin_boy_should_trash`, `use_amphora_now`. |

## Notes / known approximations

- **Stowaway's Reaction-to-Attack** half is not implemented — the card works as a plain Action-Duration. Adding it requires generalizing the Moat/Shield attack-reveal mechanism to support arbitrary reactions in hand.
- **Mirror / Deliver / Prepare / Journey events** are deferred — they need set-aside-on-event tracking that doesn't yet exist in the engine.

## Test plan

- [x] `pytest` — 401 pass, 0 fail
- [x] All 31 new cards register and instantiate via `get_card()`
- [x] All 7 new events register and instantiate via `get_event()`
- [x] Treasure-registry integrity test updated for new treasure cards (Crucible, Gondola, Pendant, Silver Mine, Rope, Cage, Mining Road)
- [x] Search and Cutthroat round-1 simplifications fixed and re-tested with the new hooks

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: dominion/cards/plunder/cabin_boy.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            if card.is_duration and card.cost.coins <= 5 and card.cost.potions == 0:
+                durations.append(card)
+
+        prefer_trash = False
+        if durations and hasattr(player.ai, "cabin_boy_should_trash"):
+            prefer_trash = bool(player.ai.cabin_boy_should_trash(game_state, player, durations))
+
+        if prefer_trash and durations:
+            if self in player.duration:
+                player.duration.remove(self)
```
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/cards/plunder/cabin_boy.py:45]: Fixed in 8e627ae. Confirmed the crash by writing a regression test that runs the trash branch through `do_duration_phase` (not just calling `on_duration` directly as the original test did). The engine then unconditionally calls `player.duration.remove(card)` and raises `ValueError`.

Fix: in the trash branch, set `duration_persistent = True` after handling the cleanup ourselves. That signals to `do_duration_phase` to skip the remove since we've already moved the card to trash.

File: dominion/cards/plunder/mining_road.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            return
+
+        choice = owner.ai.choose_buy(game_state, list(candidates) + [None])
+        if choice is None or game_state.supply.get(choice.name, 0) <= 0:
+            return
+
+        self._gain_reaction_armed = False
+
+        owner.hand.remove(chosen_treasure)
+        owner.in_play.append(chosen_treasure)
```
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/cards/plunder/mining_road.py]: Declining — this comes down to a textual interpretation that I'm reading the other way.

The canonical text is "the next time you gain a non-Victory card during your Buy phase this turn, you may **play** a Treasure card from your hand: gain a card costing exactly $1 more than that Treasure." On my reading, "play a Treasure" means moving it to in-play and resolving its on-play effects (granting its coin, etc.) — not setting it aside.

Under that reading, Silver Mine's `on_discard_from_play` triggering a Silver gain at end of turn is the **correct** behaviour: the player did play Silver Mine during the buy phase as part of resolving Mining Road. The same logic applies to other in-play-aware effects (Hoard, Talisman, etc.).

The "set aside" reading you describe (don't grant the coin, don't trigger discard hooks) would also be defensible if the rules text said "set aside" — but the language I'm working from says "play."

Happy to flip if you can point me at an authoritative ruling that resolves it the other way.

File: dominion/game/game_state.py [hunk 3]
```diff
@@ -1483,6 +1574,100 @@ def _handle_opponent_gain_hooks(self, gainer: PlayerState, gained_card: Card) ->
...
+        """Mirror event: next Action gained may be doubled."""
+
+        if not getattr(gainer, "mirror_armed", False):
+            return
+        if not gained_card.is_action:
+            return
+        if self.supply.get(gained_card.name, 0) <= 0:
+            return
+        if not gainer.ai.should_use_mirror(self, gainer, gained_card):
+            return
```
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/game/game_state.py:1891]: Real bug — fixed in 8e627ae.

Confirmed: a player who declines (`should_use_mirror` returns False) on their first Action gain after buying Mirror would still see `mirror_armed=True`, so a later Action gain in the same turn would re-prompt. That violates the "the next time you gain an Action card" scoping.

Fix: clear `mirror_armed` on the first eligible Action gain regardless of whether the copy is taken. Added a regression test that buys Mirror with a decline-AI, gains two Actions in sequence, and asserts only one copy was made and the trigger is consumed after the first gain.

File: dominion/game/game_state.py [hunk 4]
```diff
@@ -1483,6 +1574,100 @@ def _handle_opponent_gain_hooks(self, gainer: PlayerState, gained_card: Card) ->
...
+                if Counter(id(c) for c in ordered) == Counter(id(c) for c in cards):
+                    cards = ordered
+
+        for card in cards:
+            player.in_play.append(card)
+            actions_before = player.actions
+            card.on_play(self)
+            # Restore actions: Prepare lets you play actions without spending.
+            if card.is_action:
+                player.actions = actions_before
```
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/game/game_state.py]: Real bug — fixed in 8e627ae.

You're right: the prepared-cards code path never decrements `player.actions` (it bypasses the action-phase loop), so capturing-and-restoring `actions` actually wiped the +Actions that the played card itself granted. A prepared Village would net to 0 instead of +2.

Fix: drop the capture/restore. `on_play` adds whatever +Actions the card gives, and since nothing decrements before it, the bonus persists naturally. Updated `test_prepare_plays_set_aside_cards_at_start_of_next_turn` to assert that a prepared Village leaves the player at base 1 + 2 = 3 actions.

File: dominion/game/game_state.py [hunk 5]
```diff
@@ -295,10 +295,28 @@ def handle_start_phase(self):
             self.current_player.hand.extend(self.current_player.delayed_cards)
             self.current_player.delayed_cards = []
 
+        # Return any cards set aside by the Deliver event last turn
+        if self.current_player.delivered_cards:
+            self.current_player.hand.extend(self.current_player.delivered_cards)
+            self.current_player.delivered_cards = []
+        self.current_player.deliver_armed = False
```
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/game/game_state.py]: Real bug — fixed in 8e627ae.

You're right that off-turn gains (Trader, Watchtower-style reactions, Possession in principle, etc.) shouldn't be diverted by Deliver. The flag was clearing at the start of the buyer's next turn, which left it active throughout opponents' intervening turns.

Fix: clear `deliver_armed` at the end of the buyer's own cleanup phase. The `delivered_cards` mat still carries the cards forward and they return to hand at the buyer's next start phase. Added a regression test that runs cleanup, then makes an off-turn gain on the buyer and asserts it lands in `discard` rather than `delivered_cards`.

File: dominion/game/game_state.py [hunk 6]
```diff
@@ -295,10 +295,27 @@ def handle_start_phase(self):
...
+        # Return any cards set aside by the Deliver event last turn
+        if self.current_player.delivered_cards:
+            self.current_player.hand.extend(self.current_player.delivered_cards)
+            self.current_player.delivered_cards = []
+
+        # Resolve any Prepare cards set aside last turn: play them in any
+        # order, then they end up in play (and get discarded normally at
+        # cleanup).
+        if self.current_player.prepared_cards:
+            self._resolve_prepared_cards(self.current_player)
```
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/game/game_state.py]: Real bug — fixed in 97c549b.

Confirmed: prepared cards were resolved at line 307 of `handle_start_phase`, before `do_duration_phase` at the end of the same method. A prepared Longship's `on_play` adds itself to `player.duration`, then `do_duration_phase` immediately fires `on_duration`, granting the "next turn" +2 Cards on the wrong turn.

Fix: move the prepared-cards resolution to *after* `do_duration_phase`. Existing durations resolve first; any new durations added by prepared cards stay parked in `player.duration` and wait for the following turn.

Added regression test `test_prepared_duration_does_not_fire_on_same_turn` that prepares a Longship and asserts +2 Actions persist this turn but the +2 Cards effect doesn't fire yet.

File: dominion/game/game_state.py [hunk 7]
```diff
@@ -1419,21 +1499,34 @@ def _player_blocks_attack(self, player: PlayerState) -> bool:
...
+        Each card may return ``"block"`` to prevent the attack from resolving,
+        or any other value (typically ``None``) to indicate a non-blocking
+        side effect (e.g. Stowaway revealing for +1 Card).
+        """
 
-        self.log_callback(("action", player.ai.name, "reveals Moat to block the attack", {}))
-        return True
+        blocked = False
+        # Snapshot to avoid mutation issues if a card removes itself from hand.
+        for card in list(player.hand):
```
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/game/game_state.py]: Real bug — fixed in 97c549b.

You're right that the snapshot pattern misses Reactions drawn during an earlier reaction in the same attack window. The Stowaway-draws-Moat case is a legitimate Dominion interaction.

Fix: convert `_dispatch_attack_reactions` to a fixed-point loop. Each pass scans `player.hand` for Reactions whose `id()` hasn't been offered yet; if a non-blocking reaction draws another Reaction into hand, the next pass picks it up. Tracking by `id` (not by name) prevents re-prompting the same physical card while still allowing multiple Moats / Stowaways / etc.

Added regression test `test_attack_reactions_offer_cards_drawn_during_reaction` that puts Stowaway in hand and Moat on top of deck, runs an attack, and asserts the drawn Moat blocks it.

File: dominion/game/game_state.py [hunk 8]
```diff
@@ -850,6 +925,9 @@ def handle_cleanup_phase(self):
         ]
         player.highwayman_blocked_this_turn = False
         player.insignia_active = False
+        # Deliver only diverts gains during the buyer's own turn; clear the
+        # flag at the end of cleanup so opponent-turn gains aren't intercepted.
+        player.deliver_armed = False
```
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/game/game_state.py:1129]: Real bug — fixed in 7efe51b.

Confirmed: `mirror_armed` was only cleared by `_handle_mirror_reaction` on an actual Action gain. If the buyer didn't gain an Action that turn, the flag persisted indefinitely and would double a future Action gain — possibly even on a different turn or after taking damage from end-of-turn effects.

Fix: clear `mirror_armed` at the end of cleanup alongside `deliver_armed`. Added regression test that buys Mirror, runs cleanup with no Action gain, then gains a Village on the next turn and asserts only one copy lands in discard.

File: dominion/game/game_state.py [hunk 9]
```diff
@@ -1483,6 +1583,107 @@ def _handle_opponent_gain_hooks(self, gainer: PlayerState, gained_card: Card) ->
...
+                from collections import Counter
+
+                if Counter(id(c) for c in ordered) == Counter(id(c) for c in cards):
+                    cards = ordered
+
+        for card in cards:
+            player.in_play.append(card)
+            # Prepare plays don't spend an action — we never decrement
+            # ``player.actions`` here, so on_play's +Actions persist naturally.
+            card.on_play(self)
```
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/game/game_state.py:1845]: Real bug — fixed in 7efe51b.

Confirmed: `_resolve_prepared_cards` called `card.on_play(self)` directly without dispatching the on_action_played hook, so a Frigate (in any player's duration) wouldn't see prepared Action plays.

Fix: after `card.on_play(self)`, call `self._dispatch_on_action_played(player, card)` if the card is an Action. Treasures correctly skip the dispatch (no card hooks on Treasure plays anyway).

Added regression test where p1 has Frigate active in duration and p2 plays a Village from their prepared mat — Frigate forces p2 to discard down to 4.

File: dominion/cards/plunder/sailor.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+            gainer.hand.remove(gained_card)
+        else:
+            return
+
+        self._gain_play_armed = False
+        gainer.in_play.append(gained_card)
+        gained_card.on_play(game_state)
+
+    def on_discard_from_play(self, game_state, player):
+        self._gain_play_armed = False
```
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/cards/plunder/sailor.py:80]: Real bug — fixed in 7efe51b.

Confirmed the edge case is reachable: Sailor stays in `player.duration` through turn-N cleanup. On turn N+1, `handle_start_phase` runs the Shaman setup rule (line ~315) *before* `do_duration_phase` (line ~344), so a Shaman trash-gain on turn N+1 would fire `Sailor.on_card_gained` while `_gain_play_armed` is still True from turn N — auto-playing a Duration that should have been a normal gain.

Fix: track `self._armed_turn = player.turns_taken` when Sailor is played. In `on_card_gained`, reject gains where `owner.turns_taken != self._armed_turn` (and clear the flag on the way out so the early-return is cheap on subsequent gains).

Added regression test that plays Sailor on turn 1, runs cleanup, bumps `turns_taken`, gains a Crew, and asserts the Crew lands in discard rather than in_play.

File: dominion/events/plunder_events.py [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+    """$4 Event: Take an extra turn after this one, in which you don't draw
+    a new hand.
+    """
+
+    def __init__(self):
+        super().__init__("Journey", CardCost(coins=4))
+
+    def on_buy(self, game_state, player) -> None:
+        game_state.extra_turn = True
+        player.skip_next_draw_phase = True
```
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/events/plunder_events.py:237]: Real bug — partially fixed in 35f5f16.

I addressed the chain-extra-turn exploit you flagged: a player with $4 each turn could buy Journey on the extra turn it just granted, recursing forever. Added `GameState.is_extra_turn` (True iff the current turn was granted by an unconsumed `extra_turn` flag at the previous cleanup); Journey now refuses to fire when that flag is set. Regression test asserts `extra_turn` stays False if Journey is bought during an extra turn, and that it still works on a normal turn.

I'm **not** adding the Outpost-style "cards in play stay through cleanup" behavior. The canonical text I'm working from is "Take an extra turn after this one. You don't draw a new hand at the end of this turn" — no provision about preserving in-play. If you can point at the rulebook excerpt that calls for it, happy to add; otherwise I think the chain-prevention is the substantive fix here.

File: dominion/events/plunder_events.py [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+    def __init__(self):
+        super().__init__("Prepare", CardCost(coins=5))
+
+    def on_buy(self, game_state, player) -> None:
+        # Move in-play (excluding durations that are still resolving) and hand
+        # to the prepared mat. Treat durations as ineligible to set aside so
+        # their lingering effects still resolve.
+        moveable_in_play = [
+            c for c in player.in_play if c not in player.duration
+        ]
```
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/events/plunder_events.py]: Real bug — fixed in 35f5f16.

You're right. Moving in-play cards to the prepared mat stripped this-turn play state (Treasures already granting coin, set-asides, etc.) and replayed already-played cards next turn — way beyond what Prepare should do.

Fix: restrict Prepare to setting aside only the player's hand. In-play stays untouched and follows normal cleanup. Updated `test_prepare_sets_aside_in_play_and_hand` → `test_prepare_sets_aside_hand` to assert in-play cards remain in play, and added a new test that confirms the same.

File: dominion/cards/plunder/mining_road.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+        treasures = [c for c in owner.hand if c.is_treasure]
+        if not treasures:
+            return
+
+        chosen = owner.ai.mining_road_play_treasure(
+            game_state, owner, list(treasures), gained_card
+        )
+        if chosen is None or chosen not in owner.hand:
+            return
```
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/cards/plunder/mining_road.py:53]: Real bug — fixed in 35f5f16.

Same pattern as the Mirror P1 from round 3: "next time" triggers must consume on the first eligible event regardless of whether the optional effect is taken.

Fix: in `on_card_gained`, set `_gain_reaction_armed = False` immediately after the eligibility checks pass (non-Victory gain during buy phase), then attempt the optional Treasure-pay effect. Subsequent non-Victory gains in the same buy phase are no longer offered the choice.

Added regression test that arms Mining Road, makes a non-Victory gain with a declining AI, asserts the trigger is consumed, then makes another non-Victory gain and asserts the trigger doesn't re-fire.

File: dominion/cards/plunder/gondola.py [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+        actions_in_hand = [c for c in player.hand if c.is_action]
+        if not actions_in_hand:
+            return
+
+        choice = player.ai.choose_action(game_state, list(actions_in_hand) + [None])
+        if choice is None or choice not in player.hand:
+            return
+
+        if player.actions <= 0:
+            player.actions += 1
```
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/cards/plunder/gondola.py]: Real bug — fixed in 944cab8.

You're right. The synthetic `+1 Action` was a leftover from when I was worried that a 0-action player couldn't play an Action; but `card.on_play()` doesn't check `player.actions` (only the action-phase loop does), so the bump was both unnecessary and incorrect — it gave a permanent extra action when Gondola was gained at actions=0.

Fix: remove the bump entirely. The played card's own +Actions persist naturally because nothing decrements first. Regression test gains Gondola at actions=0 with a Village in hand and asserts player.actions == 2 (Village's +2), not 3.

File: dominion/cards/plunder/gondola.py [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+
+        player.hand.remove(choice)
+        player.in_play.append(choice)
+        choice.on_play(game_state)
```
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/cards/plunder/gondola.py:56]: Real bug — fixed in 944cab8.

Same pattern as the Prepare round-4 fix: any code path that calls `card.on_play()` directly (bypassing the action phase) needs to also dispatch the global on-action-played hooks for things like Frigate.

Fix: after `choice.on_play(game_state)` in `Gondola.on_gain`, call `game_state._dispatch_on_action_played(player, choice)`. Regression test puts Frigate in p1's duration, has p2 gain a Gondola while holding Village, and asserts Frigate forces p2 to discard down to 4 after the Village is played.

File: dominion/game/game_state.py [hunk 16]
```diff
@@ -1483,6 +1590,111 @@ def _handle_opponent_gain_hooks(self, gainer: PlayerState, gained_card: Card) ->
...
```
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/game/game_state.py:1845]: Real bug — fixed in 944cab8.

Confirmed: a Victory card prepared from hand would get appended to `in_play` and have its (no-op) `on_play` called, then sit in play through the turn — affecting in-play counts that other cards check (Pendant, Crucible, etc.) and never normally being discarded that way.

Fix: skip non-(Action|Treasure) cards in `_resolve_prepared_cards` and discard them directly. Regression test prepares an Estate + Silver, runs start phase, asserts Silver lands in_play and Estate lands in discard.

File: dominion/cards/plunder/gondola.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/cards/plunder/gondola.py:56]: Real bug — fixed in 944cab8.

Same pattern as the Prepare round-4 fix: any code path that calls `card.on_play()` directly (bypassing the action phase) needs to also dispatch the global on-action-played hooks for things like Frigate.

Fix: after `choice.on_play(game_state)` in `Gondola.on_gain`, call `game_state._dispatch_on_action_played(player, choice)`. Regression test puts Frigate in p1's duration, has p2 gain a Gondola while holding Village, and asserts Frigate forces p2 to discard down to 4 after the Village is played.

File: dominion/game/game_state.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/game/game_state.py:1845]: Real bug — fixed in 944cab8.

Confirmed: a Victory card prepared from hand would get appended to `in_play` and have its (no-op) `on_play` called, then sit in play through the turn — affecting in-play counts that other cards check (Pendant, Crucible, etc.) and never normally being discarded that way.

Fix: skip non-(Action|Treasure) cards in `_resolve_prepared_cards` and discard them directly. Regression test prepares an Estate + Silver, runs start phase, asserts Silver lands in_play and Estate lands in discard.

File: dominion/cards/plunder/quartermaster.py [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+        gained = game_state.gain_card(player, choice)
+
+        # Pull the gained card out of the discard/deck and onto the mat.
+        if gained in player.discard:
+            player.discard.remove(gained)
+        elif gained in player.deck:
+            player.deck.remove(gained)
+        elif gained in player.hand:
+            player.hand.remove(gained)
+        self.set_aside.append(gained)
```
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/cards/plunder/quartermaster.py]: Real bug — fixed in 4cb55aa.

Confirmed by writing a regression test that puts a Watchtower in hand, plays Quartermaster, and observes that the gained Silver lands in `state.trash` (via Watchtower's reaction) but the old code still appended that same Silver instance to `set_aside`. Result: the same `Card` object referenced from both zones, corrupting later interactions.

Fix: in `_do_choice`, only append to `set_aside` when the gained card is actually present in discard/deck/hand and was successfully removed. If it's been redirected (Watchtower trash, Trader exchange, Royal Seal topdeck, etc.), leave it alone.

File: dominion/cards/plunder/stowaway.py [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+        )
+
+        # Reactive plays don't go through the duration cycle: the card
+        # returns to the reacting player's hand at end of turn instead.
+        player.hand.remove(self)
+        player.in_play.append(self)
+        self._return_to_hand = True
+
+        game_state.draw_cards(player, 2)
+        player.actions += 1
```
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/cards/plunder/stowaway.py:57]: Real bug — fixed in 4cb55aa.

Same pattern as Gondola round 6 and Prepare round 4: any code path that calls `card.on_play()` directly needs to also dispatch `_dispatch_on_action_played` so global listeners (Frigate, Harbor Village, etc.) see the play.

Fix: after Stowaway's reaction draws and grants +1 Action, call `game_state._dispatch_on_action_played(player, self)`. Regression test puts an active Frigate in attacker's duration, has victim react with Stowaway, and asserts Frigate forces the victim to discard down to 4.

(Edge case noted: the active Frigate's hook will also fire on the original attack, so the victim ends at 4 either way; but the canonical dispatch is what matters.)

File: dominion/events/plunder_events.py [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+
+    def __init__(self):
+        super().__init__("Scrounge", CardCost(coins=4))
+
+    def on_buy(self, game_state, player) -> None:
+        estate_in_hand = next(
+            (c for c in player.hand if c.name == "Estate"), None
+        )
+
+        if estate_in_hand is not None:
```
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/events/plunder_events.py]: Real bug — fixed in 8c343bd.

You're right that my version was a different effect entirely. Rewrote to canonical: "Choose one: Trash a card from your hand; or gain an Estate from the trash. If you trashed an Estate, also gain a Duchy." Cost dropped from $4 to $3 to match canonical.

Added `scrounge_choice` AI hook that returns `"trash"` or `"gain_estate"`. Default leans toward `"trash"` if there's anything in hand. Three regression tests cover: trashing an Estate (gain Duchy), trashing a non-Estate (no Duchy), and gaining an Estate from the trash.

File: dominion/events/plunder_events.py [hunk 20]
```diff
@@ -0,0 +1,252 @@
...
+            )
+            if choice is None or game_state.supply.get(choice.name, 0) <= 0:
+                choice = actions_available[0]
+            game_state.supply[choice.name] -= 1
+            game_state.gain_card(player, choice)
+
+        for other in game_state.players:
+            if other is player:
+                continue
+            game_state.give_curse_to_player(other)
```
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/events/plunder_events.py]: Real bug — fixed in 8c343bd.

You're right. Rewrote to the canonical effect: optionally play an Attack from hand, then gain a Duchy, a Gold, a Loot, and a Province in sequence. The cursing-everyone read I had was just wrong.

The played Attack uses the existing action-play machinery (added to in_play, on_play fires, on_action_played dispatched), so any Attack legally in hand will resolve correctly with all hooks. Two regression tests: one with Witch in hand (asserts opponents get Cursed via the Attack and the four self-gains land), one with no Attack in hand (asserts the four gains still happen).

File: dominion/events/plunder_events.py [hunk 21]
```diff
@@ -0,0 +1,252 @@
...
+
+    def __init__(self):
+        super().__init__("Prosper", CardCost(coins=10))
+
+    def on_buy(self, game_state, player) -> None:
+        _gain_random_loot(game_state, player)
+        for name in ("Silver", "Gold"):
+            if game_state.supply.get(name, 0) > 0:
+                game_state.supply[name] -= 1
+                game_state.gain_card(player, get_card(name))
```
[the remark this answers is not in the document]
Comment (jss367) [reply dominion/events/plunder_events.py]: Real bug — fixed in 8c343bd.

You're right that hardcoding Silver/Gold blocked players from picking other kingdom Treasures (Loot, Hoard, Cache, etc.). Rewrote: gain a Loot first, then iterate every distinct Treasure name in the supply and let the AI choose which ones to gain (one of each maximum).

Added `prosper_choose_treasures(state, player, available) -> list[Card]` hook. Default returns Silver/Gold/Platinum (skips Coppers and junk Treasures). Implementation enforces "differently named" by tracking gained names. Two regression tests: one verifies the AI's selections land in discard while declined cards (Copper) don't, the other confirms only one of each name is gained even when the supply has many.

**Comment by jss367** [comment]
Closing this PR — superseded by #193 (and follow-up #206) which landed on main while this was in review.

## Why close

Main now has its own full Plunder implementation: 31 kingdom cards (in `dominion/cards/plunder/kingdom_cards.py`), 15 events (`dominion/events/plunder_events.py`), and Plunder Traits — all reviewed and merged. This PR is now `CONFLICTING` against main, and resolving the conflict would either revert main's reviewed Plunder code or throw out everything in this branch.

## What was here

- 31 Plunder kingdom cards + 11 events (~493 tests passing)
- 8 rounds of Codex review feedback addressed (40+ inline replies, ~30 fixes)
- Reactive infrastructure: `on_card_gained` dispatcher, `on_action_played` dispatcher, generalized `react_to_attack` with fixed-point loop, `on_discard_from_play` redirect-return values, overpay support

## Why I'm not extracting the infrastructure as a follow-up

Main already has equivalents for `react_to_attack`, `on_discard_from_play`, and overpay. The remaining novel pieces — generic `on_card_gained` and `on_action_played` dispatchers, plus the fixed-point loop — would have **no consumers** on main, since main's Plunder cards (Cutthroat, Sailor, Mining Road, etc.) deliberately use simpler "start of next turn" mechanics rather than the gain/action reactions my versions used.

If main's Plunder ever needs to expand toward more canonical reactive mechanics, the patterns and tests here are recorded in the git history of this branch for reference.
